solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -1682,6 +1687,37 @@ public DocCollection getCollectionLive(String coll) {
     }
   }
 
+  /**
+   * fetch the collection that is already cached. This may return a null if it is not already cached
+   * This is an optimization to avoid fetching state if it is not modified. this is particularly
+   * true for PRS collections where state is rarely modified
+   */
+  private DocCollection fetchCachedCollection(String coll) {
+    String collectionPath = DocCollection.getCollectionPath(coll);
+    DocCollection c = null;
+    ClusterState.CollectionRef ref = clusterState.getCollectionRef(coll);
+    if (ref == null) return null;
+    c = ref.getOrNull();
+    if (c == null) return null;
+    Stat stat = null;
+    try {
+      stat = zkClient.exists(collectionPath, null, false);
+    } catch (Exception e) {
+      return null;
+    }
+    if (stat != null) {
+      if (!c.isModified(stat.getVersion(), stat.getCversion())) {
+        // we have the latest collection state
+        return c;
+      }
+      if (c.isPerReplicaState() && c.getChildNodesVersion() < stat.getCversion()) {
+        // only PRS is modified. just fetch it and return the new collection
+        return c.copyWith(PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null));
+      }
```

Comment:
This wouldn't add the updated DocCollection to the cache, right? So any future calls to this method will continue on the same path (exists + fetch), until some other code path updates the cache, is that right?

Reply:
actually, there is another PR that is waiting to be merged #1242 . that ensures that the this update is in-place. 

I'm waiting for perf tests to come back to merge that